src/model.ts

```diff
@@ -738,14 +738,23 @@ export class GitExtension implements IGitExtension {
    * Fetch changes from a remote repository.
    *
    * @param auth - remote authentication information
+   * @param discard - whether or not to discard all changes
    * @returns promise which resolves upon fetching changes
    *
    * @throws {Git.NotInRepository} If the current path is not a Git repository
    * @throws {Git.GitResponseError} If the server response is not ok
    * @throws {ServerConnection.NetworkError} If the request cannot be made
    */
-  async pull(auth?: Git.IAuth): Promise<Git.IResultWithMessage> {
+  async pull(
+    auth?: Git.IAuth,
+    discard = false
+  ): Promise<Git.IResultWithMessage> {
     const path = await this._getPathRepository();
+
+    if (discard) {
+      await this.resetToCommit('HEAD');
+    }
+
```

Comment:
Right now there is no prompt to warn the user that it'll discard everything. 

I would like it to just call this same dialog found in FileList
https://github.com/jupyterlab/jupyterlab-git/blob/master/src/components/FileList.tsx#L209-L232

Not sure the best way to refactor this dialog so it can be called in either the commandsAndMenu, or the model.

Any ideas?

Reply:
Hey

It should be called in the command (the model should stay agnostic to the UI).

You can extract the `showDialog` call from `FileList` to its own file (in the `widgets` folder as this is not a react component). The refactored code could look like:

```ts
  async function discardAllChanges(model: IGitExtension, trans: TranslationBundle): Promise<void> {

    const result = await showDialog({
      title: trans.__('Discard all changes'),
      body: trans.__(
        'Are you sure you want to permanently discard changes to all files? This action cannot be undone.'
      ),
      buttons: [
        Dialog.cancelButton({ label: trans.__('Cancel') }),
        Dialog.warnButton({ label: trans.__('Discard') })
      ]
    });
    if (result.button.accept) {
      try {
        await model.resetToCommit();
        return Promise.resolve(true);
      } catch (reason) {
        showErrorMessage(
          this.props.trans.__('Discard all changes failed.'),
          reason
        );
        return Promise.reject(reason);
      }
    } else {
      return Promise.reject('The user refused to discard the changes.');
    }
  };
```